components/cosmetic_filters/renderer/cosmetic_filters_js_handler.cc

```diff
@@ -237,6 +173,41 @@ void CosmeticFiltersJSHandler::AddJavaScriptObjectToFrame(
   bundle_injected_ = false;
 }
 
+// If `id` is nonzero, then the same number can be later passed to
+// `UninjectStylesheet` to remove the stylesheet from the page.
+void CosmeticFiltersJSHandler::InjectStylesheet(const std::string& stylesheet,
+                                                int id) {
+  blink::WebLocalFrame* web_frame = render_frame_->GetWebFrame();
+
+  blink::WebStyleSheetKey* style_sheet_key = nullptr;
+  if (id != 0) {
+    // Prepend a Brave-specific string to avoid collisions with stylesheets
+    // injected from other sources
+    std::string key = "BraveCFRule" + std::to_string(id);
+    inserted_stylesheet_ids.insert({id, std::make_unique<blink::WebString>(
+                                            blink::WebString::FromASCII(key))});
+    style_sheet_key = inserted_stylesheet_ids.at(id).get();
+  }
+  web_frame->GetDocument().InsertStyleSheet(
+      blink::WebString::FromUTF8(stylesheet), style_sheet_key,
+      blink::WebDocument::kUserOrigin);
```

Comment:
`InsertStyleSheet` with `kUserOrigin` provides the desired "injected stylesheet" behavior as if it were added by an extension, rather than "constructed stylesheet" from `document.adoptedStyleSheets`.

It also has the added bonus of being able to pass in a custom `blink::WebString` as an identifier for each stylesheet, allowing those sheets to be removed later by `RemoveInsertedStyleSheet`. There was previously a lot of JS code to manage this around the `document.adoptedStyleSheets` API.

`InsertStyleSheet` and `RemoveInsertedStyleSheet` are exposed directly to the JS cosmetic filtering script through `injectStylesheet` and `uninjectStylesheet`.